### PR TITLE
refactor(module:date-picker): sync to antd4.0

### PR DESCRIPTION
**Pull request by wenqi73**
## PR Checklist

image
- `DecadeTable`, `YearTable`, `MonthTable`, `DateTable` extend `AbstractTable`. These are also used in Calendar module.
- `DecadeHeader`, `YearHeader`, `MonthHeader`, `DateHeader` extend `AbstractPanelHeader`.
- All of above make up the `LibPackerModule`.
- Merge `DateRangePicker` to `AbstractPicker`.
- Not extract `CalendarFooter`, `OkButton`, `TodayButton`.

Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Application (the showcase website) / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #4634

## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
```

### CODE REVIEW

**Comment by wendellhu95** [comment]
@wenqi73 Rebase and push -f? The CI failed to report.

File: components/date-picker/lib/public-api.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+/**
+ * @license
+ * Copyright Alibaba.com All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
+ */
+
+export { DecadeHeaderComponent } from './decade-header.component';
```
Comment (wendellhu95) [inline components/date-picker/lib/public-api.ts]: 这些内部实现隐藏起来会比较好。

```suggestion
- export { DecadeHeaderComponent } from './decade-header.component';
+ export { DecadeHeaderComponent as ɵDecadeHeaderComponent  } from './decade-header.component';
```

File: components/core/time/candy-date.ts [hunk 2]
```diff
@@ -42,6 +42,23 @@ export function sortRangeValue(rangeValue: CandyDate[]): CandyDate[] {
...
+
+export function normalizeRangeValue(value: CandyDate[]): CandyDate[] {
+  const [start, end] = value || [];
+  const newStart = start || new CandyDate();
+  const newEnd = end && end.isSameMonth(newStart) ? end.addMonths(1) : end || newStart.addMonths(1);
+  return [newStart, newEnd];
+}
+
+// tslint:disable-next-line:no-any
+export function cloneDate(value: any): any {
```
Comment (wendellhu95) [inline components/core/time/candy-date.ts]: I think types here could be exhausted? It  seems unnecessary to use `any` IMO.

**Comment by wendellhu95** [comment]
一些类型似乎有必要更改一下，比如[这里](https://github.com/NG-ZORRO/ng-zorro-antd/blob/1f490e96eea240d431810f475f1cfc9d25eb9185/components/calendar/nz-calendar.component.ts#L32-L33)

NzCalendarModeType ✅
NzCalendarDateTemplate ✅

**Comment by wenqi73** [comment]
> 一些类型似乎有必要更改一下，比如[这里](https://github.com/NG-ZORRO/ng-zorro-antd/blob/1f490e96eea240d431810f475f1cfc9d25eb9185/components/calendar/nz-calendar.component.ts#L32-L33)
> 
> NzCalendarMode ✅
> NzCalendarDateTemplate ✅
> 
> 趁 9.0 把一些命名不合理的 API 改过来

内部 api 我直接去掉 export 了。

**Comment by wenqi73** [comment]
> 一些类型似乎有必要更改一下，比如[这里](https://github.com/NG-ZORRO/ng-zorro-antd/blob/1f490e96eea240d431810f475f1cfc9d25eb9185/components/calendar/nz-calendar.component.ts#L32-L33)
> 
> NzCalendarMode ✅
> NzCalendarDateTemplate ✅
> 
> 趁 9.0 把一些命名不合理的 API 改过来

内部 api 我直接去掉 export 了。

File: components/calendar/demo/notice-calendar.ts [hunk 3]
```diff
@@ -60,10 +60,7 @@ import { Component } from '@angular/core';
 })
 export class NzDemoCalendarNoticeCalendarComponent {
   listDataMap = {
-    eight: [
-      { type: 'warning', content: 'This is warning event.' },
-      { type: 'success', content: 'This is usual event.' }
-    ],
+    eight: [{ type: 'warning', content: 'This is warning event.' }, { type: 'success', content: 'This is usual event.' }],
```
[the remark this answers is not in the document]
Comment (wenqi73) [reply components/calendar/demo/notice-calendar.ts]: calendar 部分我另提个 pr

File: components/date-picker/lib/abstract-table.html [hunk 4]
```diff
@@ -0,0 +1,69 @@
+<table class="ant-picker-content" cellspacing="0" role="grid">
```
[the remark this answers is not in the document]
Comment (wenqi73) [reply components/date-picker/lib/abstract-table.html]: 这个要复用的，就写在外面了。https://github.com/NG-ZORRO/ng-zorro-antd/pull/4699/files#diff-a2d0bd68dfa85fdacca5c02423f86f0fR33

**Comment by wenqi73** [comment]
@wendellhu95 Please review again.

**Comment by wendellhu95** [comment]
https://github.com/NG-ZORRO/ng-zorro-antd/issues?q=is%3Aopen+is%3Aissue+label%3A%22Component%3A+DatePicker%22 需要再 check 一下 DatePicker 现在有的 issue，如果有需要破坏性变更才能修复的问题要处理掉了

**Pull request merged**